### PR TITLE
SIM_FLUTTER-6 [BE] Create reset password API

### PR DESCRIPTION
**Pull request by CaneteJoash**
#### Backlog link
- https://framgiaph.backlog.com/board/SIM_FLUTTER?selectedIssueKey=SIM_FLUTTER-36&category=365505

#### slack link
- https://sun-philippine.slack.com/archives/C059L7XAXFB/p1687765247374069

#### Definition of done
- [x] user should be able to request forgot password
- [x] user should be able to receive reset password link in email 
- [x] user should be able to create his new password

#### Steps

**1. Set up env**

1. create `mailtrap account`
     - if you have already mailtrap account you can proceed to next step
2. Click `add Inbox`
image

3. Name it `sph-flutter` and then click `save`
image

4. Once save Click `sph-flutter` and change the `cURL` to `Laravel 7+'
image

5. update your `.env file` and add your credentials. 
- for you `MAIL_PASSWORD` you can view it in `show credentials`
image

6. Refer to my `.env file` for reference  you can copy the line found in `.env.example`
image

**2. Forgot password**
1. open postman and click the `post forgot_password` under auth collection
- input your email and click send
image

**3. Resetting password or change password**
1. open [Mailtrap](https://mailtrap.io/inboxes) and click the message  and copy the token below
image

2. paste the token in post man and change your password
- click the `reset password` under auth in the left corner then paste the copied token in the token value 
image

### CODE REVIEW

**Comment by Dbulawan13** [comment]
Comment:

- [ ] Please add a validation that prevents the user from resetting their password to be the same as their old password.

image

Comment: 1

File: backend/routes/api.php [hunk 1]
```diff
@@ -25,6 +25,9 @@
 Route::post('login', [AuthController::class, 'login']);
 Route::apiResource('auth', AuthController::class)->only('store');
 
+Route::post('forgot.password', [AuthController::class, 'forgotPassword']);
```
Comment (jstephend-sun) [inline backend/routes/api.php]: is there any specific reason as to why " . " (dot) is being used here?

this is not a resourceful route so i dont think   .   is suited to be used here for declaring the route

File: backend/app/Http/Controllers/AuthController.php [hunk 2]
```diff
@@ -69,4 +75,49 @@ public function update(UpdatePasswordRequest $request, User $user)
...
+
+    public function forgotPassword(ForgotPasswordRequest $request)
+    {
+        // Send the password reset link
+        $status = Password::sendResetLink(
+            $request->only('email')
+        );
+
+        if ($status == Password::RESET_LINK_SENT) {
+            return response()->json(['message' => 'The link was sent, please check your email']);
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/AuthController.php]: use Resource for response

File: backend/app/Http/Controllers/AuthController.php [hunk 3]
```diff
@@ -69,4 +75,49 @@ public function update(UpdatePasswordRequest $request, User $user)
...
+        } else {
+            return response()->json(['message' => trans('auth.email')], 401);
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/AuthController.php]: use resource for response

File: backend/app/Http/Controllers/AuthController.php [hunk 4]
```diff
@@ -69,4 +75,49 @@ public function update(UpdatePasswordRequest $request, User $user)
...
+        }
+
+        // Throw a validation exception if the password reset link sending failed
+        throw ValidationException::withMessages([
+            'email' => [trans($status)],
+        ]);
+    }
+
+    public function reset(ResetRequest $request)
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/AuthController.php]:
```suggestion
    public function reset?(ResetRequest $request)
```

be more descriptive for the function name

File: backend/app/Http/Controllers/AuthController.php [hunk 5]
```diff
@@ -69,4 +75,49 @@ public function update(UpdatePasswordRequest $request, User $user)
...
+    {
+        // Reset the user's password
+        $status = Password::reset(
+            $request->only('email', 'password', 'password_confirmation', 'token'),
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/AuthController.php]: any reason why is "password_confirmation" being retrieved here?

use function `validated()` to retrieve validated inputs

File: backend/app/Http/Controllers/AuthController.php [hunk 6]
```diff
@@ -69,4 +75,49 @@ public function update(UpdatePasswordRequest $request, User $user)
...
+                // Delete all existing tokens associated with the user
+                $user->tokens()->delete();
+
+                // Trigger the PasswordReset event
+                event(new PasswordReset($user));
+            }
+        );
+
+        if ($status === Password::PASSWORD_RESET) {
+            return response()->json(['message' => 'Password was reset successfully'], 201);
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/AuthController.php]: make use of Resource as response

File: backend/app/Http/Requests/Auth/ResetRequest.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace App\Http\Requests\Auth;
+
+use Illuminate\Foundation\Http\FormRequest;
+use Illuminate\Validation\Rule;
+
+class ResetRequest extends FormRequest
```
Comment (jstephend-sun) [inline backend/app/Http/Requests/Auth/ResetRequest.php]:
```suggestion
class Reset?Request extends FormRequest
```

be more descriptive

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes #31

review time: 15mins
comments: 7
iteration: 1st review

File: backend/app/Http/Controllers/AuthController.php [hunk 8]
```diff
@@ -68,4 +78,64 @@ public function update(UpdatePasswordRequest $request, User $user)
 
         return UserResource::make(['message' => 'Password updated successfully.']);
     }
+
+    public function forgotPassword(ForgotPasswordRequest $request)
+    {
+        $userEmail = $request->validated();
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/AuthController.php]:
```suggestion
        $ = $request->validated();
```
variable name seems inappropriate

File: backend/app/Http/Controllers/AuthController.php [hunk 9]
```diff
@@ -68,4 +78,64 @@ public function update(UpdatePasswordRequest $request, User $user)
...
+        $user = User::where('email', $validated['email'])->first();
+
+        if (!$user || Hash::check($validated['password'], $user->password)) {
+            throw ValidationException::withMessages([
+                'message' => 'password should not be an old password',
+            ]);
+        }
+
+        $status = Password::reset(
+            $request->only('email', 'password', 'token'),
```
Comment (jstephend-sun) [inline backend/app/Http/Controllers/AuthController.php]:
```suggestion
            $->only('email', 'password', 'token'),
```
use already extracted input data

File: backend/app/Models/User.php [hunk 10]
```diff
@@ -76,4 +77,10 @@ public function userTransactions()
     {
         return $this->hasMany(Transaction::class);
     }
+
+    public function sendPasswordResetNotification($token)
+    {
+        $url = '{{api}}/reset-password?token=' . $token;
```
Comment (jstephend-sun) [inline backend/app/Models/User.php]:
```suggestion
        $url = '?{{api}}?/reset-password?token=' . $token;
```
why are you putting the postman variable here? do we expect users to have the same postman setup?

File: backend/app/Models/User.php [hunk 11]
```diff
@@ -76,4 +77,10 @@ public function userTransactions()
...
+        $this->notify(new ResetPasswordNotification($url));
```
Comment (jstephend-sun) [inline backend/app/Models/User.php]:
```suggestion
        $this->notify(new ResetPasswordNotification($token));
```
remove above $uri
we should send directly the token, we dont have web for laravel

File: backend/app/Notifications/ResetPasswordNotification.php [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+class ResetPasswordNotification extends Notification
+{
+    use Queueable;
+
+    /**
+     * The password reset URL.
+     *
+     * @var string
+     */
+    protected $url;
```
Comment (jstephend-sun) [inline backend/app/Notifications/ResetPasswordNotification.php]: remove

File: backend/app/Notifications/ResetPasswordNotification.php [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * Create a new notification instance.
+     *
+     * @param  string  $url
+     * @return void
+     */
+    public function __construct(string $url)
+    {
+        $this->url = $url;
```
Comment (jstephend-sun) [inline backend/app/Notifications/ResetPasswordNotification.php]:
```suggestion

```
remove

File: backend/app/Notifications/ResetPasswordNotification.php [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    protected $url;
+
+    /**
+     * Create a new notification instance.
+     *
+     * @param  string  $url
+     * @return void
+     */
+    public function __construct(string $url)
```
Comment (jstephend-sun) [inline backend/app/Notifications/ResetPasswordNotification.php]:
```suggestion
    public function __construct(protected string $token)
```
immediately set class properties from constructor

File: backend/app/Notifications/ResetPasswordNotification.php [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+     * Get the mail representation of the notification.
+     *
+     * @param  mixed  $notifiable
+     * @return \Illuminate\Notifications\Messages\MailMessage
+     */
+    public function toMail($notifiable)
+    {
+        return (new MailMessage)
+                    ->line('Forgot Password?')
+                    ->action('Click to reset', $this->url)
```
Comment (jstephend-sun) [inline backend/app/Notifications/ResetPasswordNotification.php]:
```suggestion
->line("Copy reset token $this->token")
```

notify user to copy reset token

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes 

additional review time: 1hr
comments: 8
iteration: 2nd comment

**Review by Dbulawan13 — DISMISSED** [review]
Approve PR #31 

Working:
- Able to change password successfully
- Unable to change password, if password does not match 
- Unable to change password, if new password is the same as the old password
- Unable to change password if email is incorrect.






Testing Hours: 2
Comment: 1
Iteration: 1

**Review by jstephend-sun — APPROVED** [review]
Approved #31

additional review time: 1hr
comments: 8
iteration: 2nd comment

**Pull request merged**